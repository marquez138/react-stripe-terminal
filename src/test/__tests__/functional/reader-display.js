import React, { Component } from 'react';
import TestRenderer from 'react-test-renderer';
import UUID from 'pure-uuid';
import { TestSetup, TestComponent } from '../../helpers/test-setup';
import { StripeTerminalProvider, injectStripeTerminal } from '../../../index';

class LineItemComponent extends TestComponent {
    render() {
        const { stripeTerminal } = this.props;

        return (
            <div>
                <p>Basket Items</p>
                <p>Subtotal: {stripeTerminal.payment.subtotal}</p>
                <p>Tax: {stripeTerminal.payment.tax}</p>
                <p>
                    Balance Due:
                    {stripeTerminal.payment.balanceDue}
                </p>
                {stripeTerminal.payment.lineItems.map((item, index) => (
                    <div
                        id={
                            item.id !== undefined
                                ? 'id-rendered'
                                : 'id-not-rendered'
                        }
                        key={index}
                    >
                        <p>amount: {item.amount}</p>
                        <p>quantity: {item.quantity}</p>
                        <p>description: {item.description}</p>
                    </div>
                ))}
            </div>
        );
    }
}
describe('reader-display', () => {
    it('adds line items to display and computes total and tax', async () => {
        const TestApp = injectStripeTerminal(LineItemComponent);
        let stripeTerminal;
        const testElement = (
            <StripeTerminalProvider
                onFetchConnectionToken={TestSetup.FetchConnectionToken}
                onPaymentIntentRequest={() => {}}
                currency={'USD'}
                taxRate={0.07}
            >
                <TestApp
                    stripeTerminalTestHook={t => {
                        stripeTerminal = t;
                    }}
                />
            </StripeTerminalProvider>
        );
        const renderer = TestRenderer.create(testElement);
        let readers = await stripeTerminal.discoverReaders({
            method: 'simulated',
        });
        await stripeTerminal.connectReader(readers[0]);
        for (let i = 0; i < 3; i += 1) {
            await stripeTerminal.addLineItem({
                id: new UUID(4).toString(),
                description: 'Silver Hat',
                amount: 2000,
                quantity: 2,
            });
        }
        renderer.update(testElement);
        expect(renderer.toJSON()).toMatchSnapshot();
        await stripeTerminal.disconnectReader();
        renderer.unmount();
    });
    describe('removeLineItems', () => {
        it('removes items from display and state computes total and tax', async () => {
            const TestApp = injectStripeTerminal(LineItemComponent);
            let stripeTerminal;
            const testElement = (
                <StripeTerminalProvider
                    onFetchConnectionToken={TestSetup.FetchConnectionToken}
                    onPaymentIntentRequest={() => {}}
                    currency={'USD'}
                    taxRate={0.07}
                >
                    <TestApp
                        stripeTerminalTestHook={t => {
                            stripeTerminal = t;
                        }}
                    />
                </StripeTerminalProvider>
            );
            const renderer = TestRenderer.create(testElement);
            let readers = await stripeTerminal.discoverReaders({
                method: 'simulated',
            });
            await stripeTerminal.connectReader(readers[0]);
            let items = [];
            for (let i = 0; i < 3; i += 1) {
                let item = {
                    id: new UUID(4).toString(),
                    description: 'Silver Hat',
                    amount: 2000,
                    quantity: 1,
                };
                await stripeTerminal.addLineItem(item);
                items.push(item);
            }
            renderer.update(testElement);
            expect(renderer.toJSON()).toMatchSnapshot();
            for (let i = 0; i < items.length; i += 1) {
                await stripeTerminal.removeLineItem(items[i]);
                renderer.update(testElement);
                expect(renderer.toJSON()).toMatchSnapshot();
            }
            await stripeTerminal.disconnectReader();
            renderer.unmount();
        });
        it('decrements quantity from display and state computes total and tax', async () => {
            const TestApp = injectStripeTerminal(LineItemComponent);
            let stripeTerminal;
            const testElement = (
                <StripeTerminalProvider
                    onFetchConnectionToken={TestSetup.FetchConnectionToken}
                    onPaymentIntentRequest={() => {}}
                    currency={'USD'}
                    taxRate={0.07}
                >
                    <TestApp
                        stripeTerminalTestHook={t => {
                            stripeTerminal = t;
                        }}
                    />
                </StripeTerminalProvider>
            );
            const renderer = TestRenderer.create(testElement);
            let readers = await stripeTerminal.discoverReaders({
                method: 'simulated',
            });
            await stripeTerminal.connectReader(readers[0]);
            let items = [];
            for (let i = 0; i < 3; i += 1) {
                let item = {
                    id: new UUID(4).toString(),
                    description: 'Silver Hat',
                    amount: 2000,
                    quantity: 2,
                };
                await stripeTerminal.addLineItem(item);
                items.push(item);
            }
            renderer.update(testElement);
            expect(renderer.toJSON()).toMatchSnapshot();
            for (let i = 0; i < items.length; i += 1) {
                await stripeTerminal.removeLineItem(items[i]);
                renderer.update(testElement);
                expect(renderer.toJSON()).toMatchSnapshot();
            }
            await stripeTerminal.disconnectReader();
            renderer.unmount();
        });
    });
    it('computes custom total and tax for line items', async () => {
        const TestApp = injectStripeTerminal(LineItemComponent);
        let stripeTerminal;
        // test verification handler for custom subtotal and tax compute
        const verifySubtotalTaxHandler = items => {
            expect(items.length).toBeGreaterThan(0);
            items.forEach(item => {
                expect(item.amount).toBeGreaterThan(0);
                expect(typeof item.id).toBe('string');
                expect(item.quantity).toBeGreaterThan(0);
            });
            return 5;
        };
        const testElement = (
            <StripeTerminalProvider
                onFetchConnectionToken={TestSetup.FetchConnectionToken}
                onPaymentIntentRequest={() => {}}
                currency={'USD'}
                taxRate={0.07}
                computeSubtotal={verifySubtotalTaxHandler}
                computeTaxAmount={verifySubtotalTaxHandler}
            >
                <TestApp
                    stripeTerminalTestHook={t => {
                        stripeTerminal = t;
                    }}
                />
            </StripeTerminalProvider>
        );
        const renderer = TestRenderer.create(testElement);
        const readers = await stripeTerminal.discoverReaders({
            method: 'simulated',
        });
        await stripeTerminal.connectReader(readers[0]);
        for (let i = 0; i < 3; i += 1) {
            await stripeTerminal.addLineItem({
                id: new UUID(4).toString(),
                description: 'Silver Hat',
                amount: 2000,
                quantity: 2,
            });
        }
        renderer.update(testElement);
        expect(renderer.toJSON()).toMatchSnapshot();
        await stripeTerminal.clearPayment();
        renderer.update(testElement);
        expect(renderer.toJSON()).toMatchSnapshot();
        await stripeTerminal.disconnectReader();
        renderer.unmount();
    });
    it('clears payment state and reader screen', async () => {
        const TestApp = injectStripeTerminal(LineItemComponent);
        let stripeTerminal;
        const testElement = (
            <StripeTerminalProvider
                onFetchConnectionToken={TestSetup.FetchConnectionToken}
                onPaymentIntentRequest={() => {}}
                currency={'USD'}
                taxRate={0.07}
            >
                <TestApp
                    stripeTerminalTestHook={t => {
                        stripeTerminal = t;
                    }}
                />
            </StripeTerminalProvider>
        );
        const renderer = TestRenderer.create(testElement);
        let readers = await stripeTerminal.discoverReaders({
            method: 'simulated',
        });
        await stripeTerminal.connectReader(readers[0]);
        for (let i = 0; i < 3; i += 1) {
            await stripeTerminal.addLineItem({
                id: new UUID(4).toString(),
                description: 'Silver Hat',
                amount: 2000,
                quantity: 2,
            });
        }
        renderer.update(testElement);
        expect(renderer.toJSON()).toMatchSnapshot();
        await stripeTerminal.clearPayment();
        renderer.update(testElement);
        expect(renderer.toJSON()).toMatchSnapshot();
        await stripeTerminal.disconnectReader();
        renderer.unmount();
    });
});
