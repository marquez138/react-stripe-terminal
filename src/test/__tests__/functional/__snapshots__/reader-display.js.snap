// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reader-display adds line items to display and computes total and tax 1`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    12000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    12420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display clears payment state and reader screen 1`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    12000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    12420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display clears payment state and reader screen 2`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    0
  </p>
  <p>
    Tax: 
    0
  </p>
  <p>
    Balance Due:
    0
  </p>
</div>
`;

exports[`reader-display computes custom total and tax for line items 1`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    5
  </p>
  <p>
    Tax: 
    5
  </p>
  <p>
    Balance Due:
    10
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display computes custom total and tax for line items 2`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    0
  </p>
  <p>
    Tax: 
    0
  </p>
  <p>
    Balance Due:
    0
  </p>
</div>
`;

exports[`reader-display removeLineItems decrements quantity from display and state computes total and tax 1`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    12000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    12420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems decrements quantity from display and state computes total and tax 2`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    10000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    10420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems decrements quantity from display and state computes total and tax 3`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    8000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    8420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      2
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems decrements quantity from display and state computes total and tax 4`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    6000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    6420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems removes items from display and state computes total and tax 1`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    6000
  </p>
  <p>
    Tax: 
    420
  </p>
  <p>
    Balance Due:
    6420
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems removes items from display and state computes total and tax 2`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    4000
  </p>
  <p>
    Tax: 
    280
  </p>
  <p>
    Balance Due:
    4280
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems removes items from display and state computes total and tax 3`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    2000
  </p>
  <p>
    Tax: 
    140
  </p>
  <p>
    Balance Due:
    2140
  </p>
  <div
    id="id-rendered"
  >
    <p>
      amount: 
      2000
    </p>
    <p>
      quantity: 
      1
    </p>
    <p>
      description: 
      Silver Hat
    </p>
  </div>
</div>
`;

exports[`reader-display removeLineItems removes items from display and state computes total and tax 4`] = `
<div>
  <p>
    Basket Items
  </p>
  <p>
    Subtotal: 
    0
  </p>
  <p>
    Tax: 
    0
  </p>
  <p>
    Balance Due:
    0
  </p>
</div>
`;
